### PR TITLE
feat: Added role assignment support for RGs

### PR DESCRIPTION
**Pull request by DaFitRobsta**
# Overview/Summary

Added support for Role Assignment at the Resource Group(s) level. Following the same pattern as Management Group(s) and Subscription(s) bicep modules. This is to address this feature request #376. [AB#25949](https://dev.azure.com/CSUSolEng/2a9d5a5a-8998-4ad0-81c8-ef3045e4da97/_workitems/edit/25949)

## This PR fixes/adds/changes/removes

1. Adds support for role assignment at the Resource Group(s) level from policyAssignmentManagementGroup.bicep module.
2. Added two modules to support granting policy assignment ID at the resource group level.
3. Updated README.md to reflect the support for Resource Group(s) role assignment.

### Breaking Changes

No breaking change, additional feature added.

## Testing Evidence

Tested policyAssignmentManagementGroup.bicep module by passing in the new parameter, parPolicyAssignmentIdentityRoleAssignmentsResourceGroups, and it successfully granted policy assignment ID to contributor rights over two resource groups in a different subscription from the management group.

Tested both modules, roleAssignmentResourceGroupMany.bicep and roleAssignmentResourceGroup.bicep, following the same parameter file structure as the existing management group and subscription parameter files. Both deployments were successful.

> $inputObject = @{
>   DeploymentName        = 'alz-RoleAssignmentsDeployment-{0}' -f (-join (Get-Date -Format 'yyyyMMddTHHMMssffffZ')[0..63])
>   Location              = 'westus3'
>   ManagementGroupId     = 'MTB-Sandboxes'
>   TemplateFile          = "infra-as-code/bicep/modules/roleAssignments/roleAssignmentResourceGroupMany.bicep"
>   TemplateParameterFile = 'infra-as-code/bicep/modules/roleAssignments/parameters/roleAssignmentResourceGroupMany.servicePrincipal.parameters.rpl.json'
> }
> New-AzManagementGroupDeployment @inputObject
> 
> **Results**:
> DeploymentName          : alz-RoleAssignmentsDeployment-20230120T1901489305Z
> ManagementGroupId       : MTB-Sandboxes
> Location                : westus3
> ProvisioningState       : Succeeded
> Timestamp               : 1/21/2023 2:11:17 AM

> $inputObject = @{
>   DeploymentName        = 'alz-RoleAssignmentsDeployment-{0}' -f (-join (Get-Date -Format 'yyyyMMddTHHMMssffffZ')[0..63])
>   ResourceGroupName     = 'rg01'
>   TemplateFile          = "infra-as-code/bicep/modules/roleAssignments/roleAssignmentResourceGroup.bicep"
>   TemplateParameterFile = 'infra-as-code/bicep/modules/roleAssignments/parameters/roleAssignmentResourceGroup.servicePrincipal.parameters.rpl.json'
> }
> New-AzResourceGroupDeployment @$inputObject 
> 
> **Results**:
> DeploymentName          : alz-RoleAssignmentsDeployment-20230120T1801291965Z
> ResourceGroupName       : rg01
> ProvisioningState       : Succeeded
> Timestamp               : 1/21/2023 1:57:41 AM

## As part of this Pull Request I have

- [x] Read the [Contribution Guide](https://github.com/Azure/ALZ-Bicep/wiki/Contributing) and ensured this PR is compliant with the guide
- [x] Ensured the resource API versions in `.bicep` file/s I am adding/editing are using the latest API version possible
- [x] Checked for duplicate [Pull Requests](https://github.com/Azure/ALZ-Bicep/pulls)
- [x] Associated it with relevant [GitHub Issues](https://github.com/Azure/ALZ-Bicep/issues)
- [x] *(ALZ Bicep Core Team Only)* Associated it with relevant [ADO Items](https://aka.ms/alz/bicep/backlog)
- [x] Ensured my code/branch is up-to-date with the latest changes in the `main` [branch](https://github.com/Azure/ALZ-Bicep/tree/main)
- [x] Performed testing and provided evidence.
- [ ] Updated one or more of the following tests *(if required)*
  - [Unit](https://github.com/Azure/ALZ-Bicep/blob/main/.github/workflows/bicep-build-to-validate.yml)
  - [Linting](https://github.com/Azure/ALZ-Bicep/tree/main/.github/workflows)
  - [E2E (End-To-End)](https://github.com/Azure/ALZ-Bicep/blob/main/tests/pipelines/bicep-build-to-validate.yml)
  - [ValidateAzCloud (Base validation in Azure Cloud)](https://github.com/Azure/ALZ-Bicep/blob/main/tests/pipelines/base-unit-validate.yml)
  - [ValidateMcCloud (Base validation in Azure China Cloud)](https://github.com/Azure/ALZ-Bicep/blob/main/tests/pipelines/mc-base-unit-validate.yml)
- [x] Updated relevant and associated documentation (e.g. Contribution Guide, Module READMEs, Wiki Docs etc.)
- [ ] If relevant, created or updated Code Tours [here](https://github.com/Azure/ALZ-Bicep/blob/main/.vscode/tours)

### CODE REVIEW

**Review by jtracey93 — CHANGES_REQUESTED** [review]
Looking good @DaFitRobsta, just a couple of small bits. 

Do you think it may be easier to take the full resource ID for resource groups as the array entries rather than a custom input of `<sub ID>/<RG Name>`?

I'm cool with not changing it just wanted to pose the question. I can see you followed the current convention we use here like subscriptions and management groups so completely fine with leaving it as I say.

**Comment by DaFitRobsta** [comment]
> Looking good @DaFitRobsta, just a couple of small bits.
> 
> Do you think it may be easier to take the full resource ID for resource groups as the array entries rather than a custom input of `<sub ID>/<RG Name>`?
> 
> I'm cool with not changing it just wanted to pose the question. I can see you followed the current convention we use here like subscriptions and management groups so completely fine with leaving it as I say.

@jtracey93, I initially considered using the resource ID of the resource group, but when setting the scope of the module, it won't accept a resource ID. It expects a subscription ID, resource Group name in the resourceGroup( ).  To reduce the number of additional string functions, I decided to go with a custom input. I can definitely change it, but after reviewing the other modules, they are very similar in that they don't use an official resource ID. Let me know how you would like me to proceed.

File: infra-as-code/bicep/modules/roleAssignments/roleAssignmentResourceGroup.bicep [hunk 1]
```diff
@@ -0,0 +1,42 @@
+targetScope = 'resourceGroup'
```
[the remark this answers is not in the document]
Comment (DaFitRobsta) [reply infra-as-code/bicep/modules/roleAssignments/roleAssignmentResourceGroup.bicep:1]: It appears scope is required based on how policyAssignmentManagementGroup.bicep is defined. When I removed the targetscope, I got the following errors:

`C:\repos\ALZ-Bicep\infra-as-code\bicep\modules\policy\assignments\policyAssignmentManagementGroup.bicep(111,8) : Error BCP035: The specified "module" declaration is missing the following required properties: "scope".

C:\repos\ALZ-Bicep\infra-as-code\bicep\modules\policy\assignments\policyAssignmentManagementGroup.bicep(111,121) : Error BCP134: Scope "managementGroup" is not valid for this module. Permitted scopes: "resourceGroup".`

**Comment by jtracey93** [comment]
Ah set scope on the module declaration to resource group in the policy module.

Sent from my iPhone

On 24 Jan 2023, at 14:23, Robert Lightner ***@***.***> wrote:

﻿

@DaFitRobsta commented on this pull request.

________________________________

In infra-as-code/bicep/modules/roleAssignments/roleAssignmentResourceGroup.bicep<https://github.com/Azure/ALZ-Bicep/pull/434#discussion_r1085393048>:

> @@ -0,0 +1,42 @@
+targetScope = 'resourceGroup'


It appears scope is required based on how policyAssignmentManagementGroup.bicep is defined. When I removed the targetscope, I got the following errors:

`C:\repos\ALZ-Bicep\infra-as-code\bicep\modules\policy\assignments\policyAssignmentManagementGroup.bicep(111,8) : Error BCP035: The specified "module" declaration is missing the following required properties: "scope".

C:\repos\ALZ-Bicep\infra-as-code\bicep\modules\policy\assignments\policyAssignmentManagementGroup.bicep(111,121) : Error BCP134: Scope "managementGroup" is not valid for this module. Permitted scopes: "resourceGroup".`

File: infra-as-code/bicep/modules/roleAssignments/roleAssignmentResourceGroup.bicep (hunk 1 above)
[the remark this answers is not in the document]
Comment (DaFitRobsta) [reply infra-as-code/bicep/modules/roleAssignments/roleAssignmentResourceGroup.bicep]: Removed the targetScope from this module.

File: infra-as-code/bicep/modules/roleAssignments/roleAssignmentResourceGroupMany.bicep [hunk 2]
```diff
@@ -0,0 +1,35 @@
+targetScope = 'managementGroup'
```
[the remark this answers is not in the document]
Comment (DaFitRobsta) [reply infra-as-code/bicep/modules/roleAssignments/roleAssignmentResourceGroupMany.bicep:1]: Keeping this one in place based on testing and conversation you and I had.

**Review by jtracey93 — APPROVED** [review]
LGTM. running tests and will merge

**Pull request merged**